### PR TITLE
Observe connections to TCP listening servers

### PR DESCRIPTION
**Pull request by LukeButters**
[SC-55388]

# Background

To observe the async halibut rollout, we would like to know the number of polling tentacle connections are being made.

# Results

Related to OctopusDeploy/Issues#8266

## Before

We had no way of knowing how often polling tentacles were connecting/disconnecting.

## After

We now have observability on the number of connections/disconnections that are occurring.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/Support/LatestClientAndLatestServiceBuilder.cs [hunk 1]
```diff
@@ -138,6 +139,18 @@ public LatestClientAndLatestServiceBuilder WithServiceStreamFactory(IStreamFacto
             return this;
         }
 
+        public LatestClientAndLatestServiceBuilder WithServiceConnectionsObserver(IConnectionsObserver countingServiceObserver)
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/Support/LatestClientAndLatestServiceBuilder.cs:142]: nit: `countingServiceObserver`

File: source/Halibut.Tests/Transport/Observability/ConnectionObserverFixture.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+using Halibut.Tests.Support;
+using Halibut.Tests.Support.TestAttributes;
+using Halibut.Tests.Support.TestCases;
+using Halibut.Tests.TestServices.Async;
+using Halibut.TestUtils.Contracts;
+using NUnit.Framework;
+
+namespace Halibut.Tests.Transport.Observability
+{
+    public class ConnectionObserverFixture : BaseTest
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/Transport/Observability/ConnectionObserverFixture.cs:13]: nit: would tests disposing of the Service be good to add e.g. tentacle shutdown and that causes the observer to detect a close

**Review by nathanwoctopusdeploy — APPROVED** [review]
LGTM

File: source/Halibut.Tests/Support/LatestClientAndLatestServiceBuilder.cs (hunk 1 above)
Comment (sburmanoctopus) [reply source/Halibut.Tests/Support/LatestClientAndLatestServiceBuilder.cs:142]: I didn't see these nits (juggling too many PRs...).

But here is a [follow up PR](https://github.com/OctopusDeploy/Halibut/pull/493) to address them

File: source/Halibut.Tests/Transport/Observability/ConnectionObserverFixture.cs (hunk 2 above)
Comment (sburmanoctopus) [reply source/Halibut.Tests/Transport/Observability/ConnectionObserverFixture.cs:13]: I had a stab at what I thought you meant by this [here](https://github.com/OctopusDeploy/Halibut/pull/493)

The cheapest way was to just verify the closure outside the disposal of the clientAndService.

Let me know if this was not what you had in mind 🙂

**Pull request merged**